hibernate-core/src/main/java/org/hibernate/collection/internal/PersistentBag.java

```diff
@@ -196,7 +199,28 @@ public boolean equalsSnapshot(CollectionPersister persister) throws HibernateExc
 	 * @return Map of "equality" hashCode to List of objects
 	 */
 	private Map<Integer, List<Object>> groupByEqualityHash(List<Object> searchedBag, Type elementType) {
-		return searchedBag.stream().collect( Collectors.groupingBy( elementType::getHashCode ) );
+		if ( searchedBag.isEmpty() ) {
+			return Collections.emptyMap();
+		}
+		Map<Integer, List<Object>> map = new HashMap<>();
+		for ( Object o : searchedBag ) {
+			map.computeIfAbsent( nullableHashCode( o, elementType ), k -> new ArrayList<>() ).add( o );
```

Comment:
using LinkedList seems not a bad idea, btw. It would simulate the hashmap bucket implementation as much as possible and avoid ArrayList sizing dilemma.